test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java

```diff
@@ -0,0 +1,207 @@
+                System.out.println(SOCKET_CLOSED_MSG);
+            } else {
+                System.out.println(SOCKET_NOT_CLOSED_MSG);
+                throw e;
+            }
+        }
+    }
+    public static class CustomSocketFactory extends SocketFactory {
+        public static CustomSocket customSocket = new CustomSocket();
+
+        public static CustomSocketFactory getDefault() {
+            return new CustomSocketFactory();
+        }
+
+        @Override
+        public Socket createSocket() {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(String s, int timeout) {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(String host, int port, InetAddress localHost,
+                                   int localPort) {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(InetAddress host, int port) {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(InetAddress address, int port,
+                                   InetAddress localAddress, int localPort) {
+            return customSocket;
+        }
+    }
+
+    private static class CustomSocket extends Socket {
+        private int closeMethodCalled = 0;
+
+        public int closeMethodCalledCount() {
+            return closeMethodCalled;
+        }
+
+        @Override
+        public void close() throws IOException {
+            closeMethodCalled++;
+            super.close();
+        }
+    }
+
+    private static void setKeyStore() {
+        String dir = System.getProperty("test.src", ".") + File.separator;
+        System.setProperty("javax.net.ssl.keyStore", dir + "ksWithSAN");
+        System.setProperty("javax.net.ssl.keyStorePassword", "welcome1");
+        System.setProperty("javax.net.ssl.trustStore", dir + "ksWithSAN");
+        System.setProperty("javax.net.ssl.trustStorePassword", "welcome1");
+    }
+
+    static class TestServer extends Thread implements AutoCloseable {
+
+        private final ServerSocket serverSocket;
+        private final int PORT;
+        private volatile boolean exceptionThrown;
+
+        TestServer() throws IOException {
+            try {
+                SSLServerSocketFactory socketFactory = (SSLServerSocketFactory) SSLServerSocketFactory.getDefault();
+                serverSocket = socketFactory.createServerSocket(0);
+                PORT = serverSocket.getLocalPort();
+            } catch (IOException ex) {
+                throw new RuntimeException(ex);
+            }
+            setDaemon(true);
+        }
+
+        public int getPortNumber() {
+            return PORT;
+        }
+
+        public boolean isExceptionThrown() {
+            return exceptionThrown;
+        }
+
+        @Override
+        public void run() {
+            try (Socket socket = serverSocket.accept()) {
+                Thread.sleep(10000);
```

Comment:
What's the purpose of the sleep ? 
Regardless, based on the test semantics alluded in the test name, the server should never enter the read block. So is this code redundant? Or is it there just in case the accept and the handshake succeeds?

Reply:
OK, So the SSL handshake must take place within the getInputStream call rather than the accept, as such you are triggering a timeout on the SSL handshake.